spec/01-unit/016-dns_spec.lua

```diff
@@ -0,0 +1,123 @@
+-- simple debug function
+local dump = function(...)
+  print(require("pl.pretty").write({...}))
+end
+
+
+describe("DNS", function()
+  local balancer, resolver, query_func, old_new
+  local mock_records, singletons, client
+
+
+  setup(function()
+    stub(ngx, "log")
+    _G._TEST = true
+    singletons = require "kong.singletons"
+
+    singletons.cache = {
+        get = function() return {} end
+    }
+
+    singletons.dao = {}
+    singletons.dao.upstreams = {
+      find_all = function(self) return {} end
+    }
+
+    balancer = require "kong.core.balancer"
+    balancer.init()
+
+    resolver = require "resty.dns.resolver"
+    client = require "resty.dns.client"
+  end)
+
+  teardown(function()
+    _G._TEST = nil
+    if type(ngx.log) == "table" then
+      ngx.log:revert()
+    end
+  end)
+
+
+  before_each(function()
+    mock_records = {}
+
+    -- you can replace this `query_func` upvalue to spy on resolver query calls.
+    -- This default will look for a mock-record and if not found call the
+    -- original resolver
+    query_func = function(self, original_query_func, name, options)
+      if mock_records[name .. ":" .. options.qtype] then
+        return mock_records[name .. ":" .. options.qtype]
+      end
+      print("now looking for: ", name .. ":" .. options.qtype)
+      return original_query_func(self, name, options)
+    end
+  
```

Comment:
This line has some whitespaces... We finally got rid of those? Have you configured your IDE to do so? If not, please do